tests/mocha/navigation_modify_test.js

```diff
@@ -0,0 +1,328 @@
+    setup(function() {
+      Blockly.Navigation.marker_.setLocation(
+          Blockly.ASTNode.createWorkspaceNode(
+              this.workspace, new goog.math.Coordinate(100, 200)));
+    });
+    test('Cursor on row block', function() {
+      Blockly.Navigation.cursor_.setLocation(
+          Blockly.ASTNode.createBlockNode(
+              this.row_block_1));
+      chai.assert.isTrue(Blockly.Navigation.modify());
+      var pos = this.row_block_1.getRelativeToSurfaceXY();
+      chai.assert.equal(100, pos.x);
+      chai.assert.equal(200, pos.y);
+    });
+
+    test('Cursor on connection', function() {
+      // TODO: Does the type of the connection matter?
+      Blockly.Navigation.cursor_.setLocation(
+          Blockly.ASTNode.createConnectionNode(
+              this.row_block_1.outputConnection));
+      chai.assert.isTrue(Blockly.Navigation.modify());
+      var pos = this.row_block_1.getRelativeToSurfaceXY();
+      chai.assert.equal(100, pos.x);
+      chai.assert.equal(200, pos.y);
+    });
+
+    test('Cursor on child block (row)', function() {
+      this.row_block_1.inputList[0].connection.connect(
+          this.row_block_2.outputConnection);
+
+      Blockly.Navigation.cursor_.setLocation(
+          Blockly.ASTNode.createBlockNode(
+              this.row_block_2));
+
+      chai.assert.isTrue(Blockly.Navigation.modify());
+      chai.assert.isNull(this.row_block_2.getParent());
+    });
+
+    test('Cursor on child block (stack)', function() {
+      this.stack_block_1.nextConnection.connect(
+          this.stack_block_2.previousConnection);
+
+      Blockly.Navigation.cursor_.setLocation(
+          Blockly.ASTNode.createBlockNode(
+              this.stack_block_2));
+
+      chai.assert.isTrue(Blockly.Navigation.modify());
+      chai.assert.isNull(this.stack_block_2.getParent());
+    });
+
+    test('Cursor on workspace', function() {
+      Blockly.Navigation.cursor_.setLocation(
+          Blockly.ASTNode.createWorkspaceNode(
+              this.workspace, new goog.math.Coordinate(100, 100)));
+      chai.assert.isFalse(Blockly.Navigation.modify());
+    });
+  });
+
+  suite('Marked Block', function() {
+    // TODO: Decide whether it ever makes sense to mark a block, and what to do
+    // if so.  For now all of these attempted modifications will fail.
+    suite('Marked any block', function() {
+      // These tests are using a stack block, but do not depend on the type of
+      // the block.
+      setup(function() {
+        Blockly.Navigation.marker_.setLocation(
+            Blockly.ASTNode.createBlockNode(
+                this.stack_block_1));
+      });
+      test('Cursor on workspace', function() {
+        Blockly.Navigation.cursor_.setLocation(
+            Blockly.ASTNode.createWorkspaceNode(
+                this.workspace, new goog.math.Coordinate(100, 100)));
+        chai.assert.isFalse(Blockly.Navigation.modify());
+
+      });
+    });
+    suite('Marked stack block', function() {
```

Comment:
Just a note: This and the 'Marked row block' suites are behavior I think we'll want to hide higher up. We should try to do something reasonable if the user marks a block and tries to insert another block.

Reply:
Agreed.  These are here partly because I was just walking through a matrix of cursor/marker combinations, and partly to point out that even if we don't want to do anything here, we need some way to signal that if it's at all possible for the user to get here.